### ISSUE TITLE
Improve structure in `NuguClient`

### ISSUE THREAD

**Issue by yHoonyK** [nugu-developers/nugu-ios#70]
# Overviews

## Category

- [ ] New feature 
- [x] Update feature
- [x] Update documentation
- [x] Idea
- [x] Architecture
- [x] Structure

## Summary
`NuguClient.Builder()` is always make all of capability-agent.
Some capability-agent may not be used. (like `LocationAgent`)

## Reference
Discussion(Internal): https://github.com/orgs/nugu-developers/teams/nugu-sdk/discussions/7

**Comment by yHoonyK** [comment]
It contains updating protocol for capability-agent.
It is that about internal mandatory property.

**Comment by yHoonyK** [comment]
```swift
public protocol ExtensionAgentProtocol:
CapabilityAgentable,
ContextInfoDelegate,
HandleDirectiveDelegate {
    /// The object that acts as the delegate of extension-agent
    var delegate: ExtensionAgentDelegate? { get set }
}
```
```swift
public protocol ExtensionAgentConcretable {
    var messageSender: MessageSendable! { get set }
}
```
```swift
final public class ExtensionAgent: ExtensionAgentProtocol, ExtensionAgentConcretable {}
```

If we want to use it like that, the coordinating role must be played by the `NuguClient.Builder`.

**Comment by yHoonyK** [comment]
Apple blog: https://developer.apple.com/swift/blog/?id=11

**Comment by yHoonyK** [comment]
discussion(internal): https://github.com/orgs/nugu-developers/teams/nugu-sdk-ios-team/discussions/3

**Comment by Minddol** [comment]
## Suggestion - Use protocol initializer instead of property.
```swift
public protocol ExtensionAgentProtocol:
CapabilityAgentable,
ContextInfoDelegate,
HandleDirectiveDelegate {
    init(messageSender: MessageSendable)
    
    /// The object that acts as the delegate of extension-agent
    var delegate: ExtensionAgentDelegate? { get set }
}
```
```swift
final public class ExtensionAgent: ExtensionAgentProtocol {
    private let messageSender: MessageSendable
    
    public init(messageSender: MessageSendable) {
        self.messageSender = messageSender
    }
}
```
```swift
    let client: NuguClient = {
        let builder = NuguClient.Builder()
        return builder
            .with(extensionAgent: ExtensionAgent(messageSender: builder.networkManager))
            .build()
    }()
```

**Comment by yHoonyK** [comment]
I think @Minddol's suggestion is the best way.

**Comment by yHoonyK** [comment]
## Proposal
Improve issue that `NuguClient.Builder()` is always make all of capability-agent.

## Concept
```swift
enum ClientBuilderDefaultType {
  case .nothing // no has default capability-agent instance
  case .basic // has basic capability-agent instances (ASRAgent, TTSAgent, TextAgent..?)
  case .complete // has all capability-agent instances
}
```

```swift
class NuguClient.Builder {
  init(defaultType: ClientBuilderDefaultType) {}
}
```

```swift
let client = NuguClient.Builder(defaultType: .basic).build()
```

**Comment by yHoonyK** [comment]
Builder pattern with director.
https://johngrib.github.io/wiki/builder-pattern/

**Comment by yHoonyK** [comment]
I consider to change architecture to 'abstract factory pattern' for capability-agent.

**Comment by yHoonyK** [comment]
#### Check-list
- [ ] Implements
- [ ] Rename
- [ ] Documentation (readme, developers .. )

**Comment by yHoonyK** [comment]
Revert renaming capability-agents.

**Comment by yHoonyK** [comment]
### Summary
#### Issue
1. `NuguClient.Builder()` is always make all of capability-agent.
   - Resolved: To inject capability-agent has be changed design-pattern to 'Abstract Factory Pattern'
2. capability-agent's access-control issue. (public properties in protocol)
   - Resolved: Remove properties to inject dependency using 'setter injection' in capability-agent protocol,   concrete class has inject dependency using 'constructor injection'.

#### Result
- `NuguClient` conform to DI(Dependency injection) pattern and using 'constructor injection' type of DI.
- `CapabilityAgentFactory` and `MediaPlayerFactory` conform to 'Abstract Factory' pattern.

**Comment by yHoonyK** [comment]
Resolved by #147

**Comment by yHoonyK** [comment]
It wasn't be completed.

**Comment by yHoonyK** [comment]
Improve capability-agent protocol structure to PoP (Protocol Oriented Programming)

**Comment by yHoonyK** [comment]
Resolved by #170 

After that, we consider that changing design-pattern to IoC container.
It will makes another github issue.
